client/extensions/woocommerce/state/ui/products/edits-reducer.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import { isNumber } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	WOOCOMMERCE_EDIT_PRODUCT,
+} from '../../action-types';
+
+const initialState = null;
+
+export default function( state = initialState, action ) {
+	const handlers = {
+		[ WOOCOMMERCE_EDIT_PRODUCT ]: editProductAction,
+	};
+
+	const handler = handlers[ action.type ];
+
+	return ( handler && handler( state, action ) ) || state;
+}
+
+function editProductAction( edits, action ) {
+	const { product, newProductIndex, data } = action.payload;
+
+	return editProduct( edits, product, newProductIndex, data );
+}
+
+function editProduct( edits, product, newProductIndex, data ) {
+	if ( newProductIndex !== undefined ) {
```

Comment:
Your action `editNewProduct` as defined in `actions.js` handles `newProductIndex` property. So presumably, it will be defined. This if statement will be falsy in the case of being called via this `editNewProduct` action. The result will be a function call to `editExistingProduct` instead of `editNewProduct`. Given the nomenclature, one would expect the opposite.

Another issue is that `editNewProduct` action can be used for each case, depending on if you define `newProductIndex`. Why not create separate events `WOOCOMMERCE_EDIT_PRODUCT` and `WOOCOMMERCE_EDIT_NEW_PRODUCT`? I think it would improve readability and avoid the dual nature of `editNewProduct` action.